### ISSUE TITLE
["BUG"] Use already installed yq before downloading a new one? | [mod=yq] [v0.3.14]

### ISSUE THREAD

**Issue by iandol** [x-cmd/x-cmd#30]
**Describe the bug**
I already have some tools installed using homebrew, including `/opt/homebrew/bin/yq`. But when I run a command like `x chat help` (maybe wrong syntax?) then x-cmd redownloads yq, which is slow and delays the command answer:

```bash
Seneca » ~ via 🐏 12GiB/16GiB | 1GiB/2GiB 
ε ➪ which yq
/opt/homebrew/bin/yq

ε ➪ x chat help
- I|pkg: Trying download ==> 'yq v4.44.2'
- I|pkg: |
    download yq=v4.44.2, thanks to the courtesy of https://registry.npmmirror.com
  more:
    url: https://registry.npmmirror.com/@x-cmd-pkg/yq/-/yq-0.0.0-v4.44.2-darwin.arm64.0.tgz
    ball: /Users/ian/.x-cmd.root/local/data/pkg/download/yq/v4.44.2.darwin.arm64.0.tgz
    name: yq
    version: v4.44.2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   261    0     0    0     0      0      0 --:--:--  0:00:05 --:--:--     0
######################################################################## 100.0%-=#=#   #   #                                     ######################################################################## 100.0%
- I|pkg: Trying populate ==> 'yq v4.44.2'
- I|pkg: Unpacking
  more:
    ball: /Users/ian/.x-cmd.root/local/data/pkg/download/yq/v4.44.2.darwin.arm64.0.tar.xz
    target_dir: /Users/ian/.x-cmd.root/local/data/pkg/sphere/X/tree.darwin.arm64.0/yq/v4.44.2
- I|pkg: Unpack is done
- I|pkg: Running /Users/ian/.x-cmd.root/local/data/pkg/metadata/v0.1.1/pkg/data/yq/.x-cmd/populate.sh
- I|x_cmd_pkg: shim gen
  more:
    source: /Users/ian/.x-cmd.root/local/data/pkg/sphere/X/tree.darwin.arm64.0/yq/v4.44.2/bin/
    shim_bin: /Users/ian/.x-cmd.root/local/data/pkg/sphere/X/tree.darwin.arm64.0/yq/v4.44.2/shim_bin/
    adaptive_shim_bin: /Users/ian/.x-cmd.root/local/data/pkg/sphere/X/tree.darwin.arm64.0/yq/v4.44.2/adaptive_shim_bin/
- I|pkg: yq v4.44.2 install successfully
- I|openai: Requesting OpenAI [model=gpt-4o]

     tell me how I can help! I need more information to be useful. 
    
    For example, you could ask me:
    
    ● To answer a question: "What is the capital of France?"
    ● To generate creative content: "Write a short story about a cat who goes on an adventure."
    ● To translate something: "Translate 'hello world' into Spanish."
    ● To summarize text: "Summarize the plot of the movie Inception."
    
    The more specific you are, the better I can assist you. 😊
    
                                                                                                                                 
Seneca » ~ via 🐏 14GiB/16GiB | 1GiB/2GiB took 26s 
ε ➪ x env ls
- E|pick: |
    The data is empty
```                                                                                                              

And it seems `yq` is just installed temporarily, if I run another command in a new terminal `yq` gets downloaded again. `x env ls` shows no `yq` installed.

**System (please complete the following information):**
 - Version: 6446d9f392c9fe23ec9a7e6e3c68cb5cba16a44ba914609fdd818ead2ccb6b2ec9c6c35dc91b18f488fb78c3528afbe9d7ccce2de355095df062b7d50f93eb0f
 - Shell: Elvish
 - OS: Darwin
 - Arch: arm64

**Screenshots**

**To Reproduce**
1. run an x command that depends on `yq`

**Expected behavior**
Search the path of the current shell to find `yq`. I assume the problem is you are running in a shell inside elvish with just a basic path, thus you do not have access to homebrew. I wonder if you could add possible package manager paths, i.e. `/opt/homebrew/bin` / `/usr/local/bin` / `~/bin` / `~/.local/bin` etc. by default if present?

**Additional context**

**Comment by edwinjhlee** [comment]
This problem is quite complicated ...

x-cmd modules rely on external binaries, like yq. Take yq as example, it behaves differently in different versions.

Our initial focus is on ensuring functionality. Once achieved, we will explore optimizations like local binary reuse. But it takes a lot of work to test.

We will first resue the big binaries (size > 50MB) first, like python.  

We consider this an optimization, not a bug.

**Comment by edwinjhlee** [comment]
```
And it seems yq is just installed temporarily, if I run another command in a new terminal yq gets downloaded again. x env ls shows no yq installed.
```

When a module ( like mod A ) requires a binary (pkg B), pkg will be downloaded, unpacked and setup in this current session. We won't install it globally.

So when you open an another session to use module A, the A still need to try to download/unpacked/setup. However, because it is already downloaded, unpacked locally ( existed in the folder we called pkg sphere ), the only thing need to do is to setup in the new session, which is just a simple path finding process, it doesn't significant increase the startup time.

The reason not to install globally is because the collision. The versions of software/libraries will affect each other. If we secrectly activate pkg B when using mod A, it will increase a lot of implicit impacts on the system.

**Comment by edwinjhlee** [comment]
```
And it seems yq is just installed temporarily, if I run another command in a new terminal yq gets downloaded again. x env ls shows no yq installed.
```

It is very strange the our pkg will installed twice... It will setup the environment ( find the yq bin path) in each local sessions. But it won't download and unpack twices. Because it is stored in the pkg sphere ( unactivated ). Unless user explicitly trigger 'x env gc yq' to save disk space.

**Comment by edwinjhlee** [comment]
But I am sure there must be a reason to cause the confusion. And there might be chances that the original deisgn doesn't implement well in some situations. We need to more informations. And we will consider providing more comprehensive log to make this process more clear.

x-cmd is design to minimum impact on the systems to avoid introducing problems ( to the maintainers, developers ) implictly . This complexity of this whole design/usage is the price. The only thing we could do is to provide better logs and articles for different use cases, and hopefully users can find the best way to use pkg as needed.

**Comment by iandol** [comment]
OK, if you consider using the `homebrew` `yq` as an optimisation, then perhaps we should close this issue?

**Comment by edwinjhlee** [comment]
> OK, if you consider using the `homebrew` `yq` as an optimisation, then perhaps we should close this issue?

I think we could leave it open. It is still valuable for us to consider.


Also can you reproduce the case where the pkg yq are downloaded twice ? That will definitely a bug which we would like to address.

**Comment by edwinjhlee** [comment]
> OK, if you consider using the `homebrew` `yq` as an optimisation, then perhaps we should close this issue?

I think it is time to evaulate the status of yq project again. Once it becomes stable, we will introduce a yq module as wrapper.

There will still be a chance that yq module is unable to use system yq if the system yq's version doesn't meet our criteria.

**Comment by lunrenyi** [comment]
## x env ls shows yq not installed
`x env ls` by default lists the packages of the current environment `x env try/use`. If you want to see the installed packages, you need to run the `x env ls -i` command.

The `x chat` module uses the `x yq` invocation form to avoid potentially affecting the user's current environment, which is the xbin invocation method.
Unlike `x env try/use`, it generates a `___x_cmd_yq` function to call the yq binary file, as follows:
```sh
___x_cmd_yq () {
  /Users/liaoxuanbin/.x-cmd.root/local/data/pkg/sphere/X/tree.darwin.arm64.0/yq/v4.44.2/bin/yq "$@"
}
```

**Comment by edwinjhlee** [comment]
This issue is resolved in v0.3.14.
Feel free to confirm and close it.
